Update ProcessLauncher::m_client to use a CheckedPtr

cdumez opened the pull request:
#### fe25acb64238f91a3880ea58c1b9755d5554d08a
<pre>
Update ProcessLauncher::m_client to use a CheckedPtr
<a href="https://bugs.webkit.org/show_bug.cgi?id=280246">https://bugs.webkit.org/show_bug.cgi?id=280246</a>

Reviewed by Darin Adler.

* Source/WebKit/UIProcess/AuxiliaryProcessProxy.h:
* Source/WebKit/UIProcess/Launcher/ProcessLauncher.h:
* Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
(WebKit::ProcessLauncher::launchProcess):

Canonical link: <a href="https://commits.webkit.org/284169@main">https://commits.webkit.org/284169@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/197da4bb375c10f903d8b7fe06a492e13b93b992

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/68569 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/47961 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/21228 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/72638 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/19714 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/55757 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/19530 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/54691 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/13108 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/71636 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/43832 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/59205 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/35154 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/40499 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/18071 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/62452 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/16975 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/74332 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/12540 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/16229 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/62160 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/12580 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/59284 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/62185 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/10127 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/3745 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10457 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/43762 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/44836 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/46030 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/44578 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebKit/UIProcess/AuxiliaryProcessProxy.h:
```diff
@@ -78,6 +78,12 @@ class AuxiliaryProcessProxy
     using ResponsivenessTimer::Client::WeakPtrImplType;
 
+    // ProcessLauncher::Client
+    uint32_t ptrCount() const final { return IPC::Connection::Client::ptrCount(); }
+    uint32_t ptrCountWithoutThreadCheck() const final { return IPC::Connection::Client::ptrCountWithoutThreadCheck(); }
+    void incrementPtrCount() const final { IPC::Connection::Client::incrementPtrCount(); }
+    void decrementPtrCount() const final { IPC::Connection::Client::decrementPtrCount(); }
```
Can this be achieved with `using` instead?

cdumez replied:
The alternative would be to have `ProcessLauncher::Client` subclass `CanMakeCheckedPtr`, and have `AuxiliaryProcessProxy` use `using` statements to disambiguate since it would inherit 2 checked ptr counts from 2 different base classes.

I can go this way if you'd prefer but I thought it was nicer to avoid inheriting 2 different checked ptr counts (cc @geoffreygaren)

geoffreygaren replied:
I think we want to avoid multiple checked counts. It's an unclear programming model.

My goal is to get to the point where every class defines its lifetime semantics in exactly one place, in a concrete class. 

Side point: IPC::Connection::Client should probably make its CheckedPtr interface virtual, just like ProcessLauncher::Client. In my view, it's an anti-pattern when an abstract base class mandates a concrete lifetime implementation. 

Another side point: RefPtr+WeakPtr is better than CheckedPtr because it guarantees correct behavior, rather than just protecting against memory un-safety with a runtime crash.

geoffreygaren replied:
Maybe this approach can help? https://godbolt.org/z/MqaT4GoP6

cdumez committed:
Update ProcessLauncher::m_client to use a CheckedPtr
https://bugs.webkit.org/show_bug.cgi?id=280246

Reviewed by Darin Adler.

* Source/WebKit/UIProcess/AuxiliaryProcessProxy.h:
* Source/WebKit/UIProcess/Launcher/ProcessLauncher.h:
* Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
(WebKit::ProcessLauncher::launchProcess):

Canonical link: https://commits.webkit.org/284169@main
